pkg/kubelet/cm/cpumanager/cpu_manager.go

```diff
@@ -206,72 +227,123 @@ func (m *manager) State() state.Reader {
-						glog.Errorf("[cpumanager] reconcileState: failed to add container (pod: %s, container: %s, container id: %s, error: %v)", pod.Name, container.Name, containerID, err)
-						failure = append(failure, reconciledContainer{pod.Name, container.Name, containerID})
-					}
-				} else {
-					// if DeletionTimestamp is set, pod has already been removed from state
-					// skip the pod/container since it's not running and will be deleted soon
-					continue
-				}
+		if podStatus.Phase == v1.PodRunning && pod.DeletionTimestamp != nil {
+			// Pod's containers have already been removed from state
+			// Skip the whole pod since it's not running and will be deleted soon
+			glog.V(5).Infof("[cpumanager] reconcile: skipping pod since deletion timestamp is set (pod: %s)", pod.Name)
+			continue
+		}
+
+		// Process containers in this pod.
+		for _, container := range containers {
+			// Skip unnamed containers (e.g. pause)
+			if container.Name == "" {
+				continue
 			}
 
-			cset := m.state.GetCPUSetOrDefault(containerID)
-			if cset.IsEmpty() {
-				// NOTE: This should not happen outside of tests.
-				glog.Infof("[cpumanager] reconcileState: skipping container; assigned cpuset is empty (pod: %s, container: %s)", pod.Name, container.Name)
-				failure = append(failure, reconciledContainer{pod.Name, container.Name, containerID})
+			// Look up container ID in the pod status. We need this in order to
+			// both get the assigned CPU set from the state and also ask the
+			// CRI to update the container resource config. The most likely
+			// cause of missing container ID is that it is not yet running.
+			containerID, err := findContainerIDByName(&podStatus, container.Name)
+			if err != nil {
+				glog.V(5).Infof("[cpumanager] reconcile: skipping container; ID not found in status (pod: %s, container: %s, error: %v)", pod.Name, container.Name, err)
+				addFailed(pod, container)
 				continue
 			}
 
-			glog.V(4).Infof("[cpumanager] reconcileState: updating container (pod: %s, container: %s, container id: %s, cpuset: \"%v\")", pod.Name, container.Name, containerID, cset)
+			// Update the container with the appropriate CPU based on the
+			// current state.
+			cset := m.state.GetCPUSetOrDefault(containerID)
+			glog.V(4).Infof("[cpumanager] reconcile: updating container (pod: %s, container: %s, container id: %s, cpuset: \"%v\")", pod.Name, container.Name, containerID, cset)
 			err = m.updateContainerCPUSet(containerID, cset)
 			if err != nil {
-				glog.Errorf("[cpumanager] reconcileState: failed to update container (pod: %s, container: %s, container id: %s, cpuset: \"%v\", error: %v)", pod.Name, container.Name, containerID, cset, err)
-				failure = append(failure, reconciledContainer{pod.Name, container.Name, containerID})
+				glog.Errorf("[cpumanager] reconcile: failed to update container (pod: %s, container: %s, container id: %s, cpuset: \"%v\", error: %v)", pod.Name, container.Name, containerID, cset, err)
+				addFailed(pod, container)
```

Comment:
If this failed once, will it ever succeed in the future?

Reply:
It's hard to know, but we won't stop trying. This involves a gRPC call to the CRI implementation. There are a number of recoverable failure scenarios I can think of.